### PR TITLE
[FIX] change creation of region_names and add test

### LINKED ISSUE #4359: [BUG] issue with masker.region_names_
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Operating system

- [ ] Linux
- [X] Mac
- [ ] Windows

### Operating system version

Mac OS Version 13.4 Ventura

### Python version

- [ ] 3.12
- [ ] 3.11
- [X] 3.10
- [ ] 3.9
- [ ] 3.8

### nilearn version

0.10.3

### Expected behavior

the output of `print(masker.region_names_)` in the following code must be `{0: 'Frontal Pole', 1: 'Insular Cortex', 2: 'Inferior Frontal Gyrus, pars triangularis'}`, because the region id 5 corresponds to region name "Inferior Frontal Gyrus, pars triangularis" based on the output of `print(initial_region_id_names)`.

The issue with region names was addressed in #4289 (this was mostly due to non-standard atlas outputs.), but it seems the issue is a bit more complicated. 

### Current behavior & error messages

This is what I got instead:

```bash
# Paste the error message here
{0: 'Frontal Pole', 1: 'Insular Cortex', 2: 'Superior Frontal Gyrus'}

```

### Steps and code to reproduce bug

```python
# Paste your code here
import numpy as np
from nilearn.regions.tests.test_signal_extraction import get_data
from nilearn import datasets
from nilearn.maskers import NiftiLabelsMasker
from nibabel import Nifti1Image

# func data
dataset = datasets.fetch_development_fmri(n_subjects=1)
func_filename = dataset.func[0]

# atlas data
atlas = datasets.fetch_atlas_harvard_oxford('cort-maxprob-thr25-2mm')

# create a mask with 3 regions
labels_data = get_data(atlas.maps)
mask_data = (labels_data == 1) + (labels_data == 2) + (labels_data == 5)
mask_img = Nifti1Image(mask_data.astype(np.int8), atlas.maps.affine)

# Instantiate the masker with label image and label values
masker = NiftiLabelsMasker(atlas.maps,
                           labels=atlas.labels,
                           mask_img=mask_img,
                           standardize=True,
                           keep_masked_labels=False
                           )

# fit and transform the masker
_ = masker.fit_transform(func_filename)

# get the region ids and names before and after masking
initial_region_ids = np.unique(labels_data)
initial_region_id_names = {region_id: atlas.labels[i] for i, region_id in enumerate(initial_region_ids)}
print(initial_region_id_names)

region_ids_after_masking = masker.region_ids_
region_names_after_masking = masker.region_names_
print(masker.region_ids_)
print(masker.region_names_)

```

### PR DESCRIPTION
**Pull request by mtorabi59**
- Closes #4359 

Changes proposed in this pull request:

- change how `region_names_` attribute of `NiftiLabelsMasker` is defined
- add tests to check the correctness of the correspondence between region names and region ids after fit_transform, e.g. resampling and masking.

### CODE REVIEW

**Comment by mtorabi59** [comment]
I am not sure if I should keep definition of `self.region_id_name` in `__init__` or move it to `fit` after definition of `self.labels_img_`

File: nilearn/maskers/tests/test_nifti_labels_masker.py [hunk 1]
```diff
@@ -841,6 +841,107 @@ def check_region_names_after_fit(
...
+        masker, region_names, region_ids, background
+    )
+
+
+def check_region_names_ids_match_after_fit(
+    masker, region_names, region_ids, background
+):
+    """Check the region names and ids correspondence.
+
+    Check that the same region names and ids correspond to eac other
```
Comment (bthirion) [inline nilearn/maskers/tests/test_nifti_labels_masker.py:928]:
```suggestion
    Check that the same region names and ids correspond to each other
```

**Review by bthirion — COMMENTED** [review]
LGTM otherwise.

**Comment by Remi-Gau** [comment]
This seems top break the report builder script in the doc build:

```
Traceback (most recent call last):
  File "/home/runner/work/nilearn/nilearn/doc/visual_testing/reporter_visual_inspection_suite.py", line 438, in <module>
    report_nifti_labels_masker()
  File "/home/runner/work/nilearn/nilearn/doc/visual_testing/reporter_visual_inspection_suite.py", line 313, in report_nifti_labels_masker
    masker = NiftiLabelsMasker(
             ^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/nilearn/maskers/nifti_labels_masker.py", line 212, in __init__
    _utils.niimg.safe_get_data(labels_img)
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/nilearn/_utils/niimg.py", line 58, in safe_get_data
    data = _get_data(img)
           ^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/site-packages/nilearn/_utils/niimg.py", line 23, in _get_data
    if img._data_cache is not None:
       ^^^^^^^^^^^^^^^
AttributeError: 'str' object has no attribute '_data_cache'
```

Can you try to run `doc/visual_testing/reporter_visual_inspection_suite.py` to see what's going on?

**Comment by mtorabi59** [comment]
Do I need to do anything regarding this PR? I think all the issues are addressed.

**Review by Remi-Gau — CHANGES_REQUESTED** [review]
Since this is a new public attribute, do you think you could update the doc string to explain what it is?

**Comment by mtorabi59** [comment]
> Since this is a new public attribute, do you think you could update the doc string to explain what it is?

I'm not sure if we want `region_id_name_` to be a public attribute, do we? My goal was to create so we can use it to create `region_names_`

**Comment by mtorabi59** [comment]
@Remi-Gau but I think it's a good idea to add some comments to explain to the developers what this attribute is

**Comment by Remi-Gau** [comment]
> > Since this is a new public attribute, do you think you could update the doc string to explain what it is?
> 
> I'm not sure if we want `region_id_name_` to be a public attribute, do we? My goal was to create so we can use it to create `region_names_`

OK in this case rename it to `_region_id_name` to signify this a "private" attribute.

**Review by bthirion — COMMENTED** [review]
Still LGTM. Probably deserves a whatsnew entry.

**Review by bthirion — APPROVED** [review]
Thx.

File: nilearn/maskers/tests/test_nifti_labels_masker.py [hunk 2]
```diff
@@ -841,6 +841,107 @@ def check_region_names_after_fit(
...
+
+    masker = NiftiLabelsMasker(
+        labels_img,
+        labels=region_names,
+        resampling_target="data",
+        mask_img=mask_img,
+        keep_masked_labels=keep_masked_labels,
+    )
+
+    _ = masker.fit().transform(fmri_img)
```
Comment (Remi-Gau) [inline nilearn/maskers/tests/test_nifti_labels_masker.py:916]:
```suggestion
    masker.fit().transform(fmri_img)
```

File: nilearn/maskers/nifti_labels_masker.py [hunk 3]
```diff
@@ -504,6 +504,41 @@ def fit(self, imgs=None, y=None):
         msg = f"loading data from {repr}"
         _utils.logger.log(msg=msg, verbose=self.verbose)
         self.labels_img_ = _utils.check_niimg_3d(self.labels_img)
+
+        # create _region_id_name dictionary
+        # this dictionary will be used to store region names and
+        # the corresponding region ids as keys
+        self._region_id_name = None
```
Comment (Remi-Gau) [inline nilearn/maskers/nifti_labels_masker.py:511]:
```suggestion
        self._region_id_name = {}
```
this way we know this is always a dict

**Review by Remi-Gau — APPROVED** [review]

**Comment by Remi-Gau** [comment]
Sorry it took me so long to get back to this one @mtorabi59

**Pull request merged**